### PR TITLE
[Compression] evaluator - step2

### PR DESCRIPTION
**Pull request by J-shang**
### Description ###
what this pr mainly done:
1. let all pruner using `Evaluator`, to make pruning lightning model easily.
2. keep the old & new API by `@overload`, using `evaluator` replace `trainer`, `traced_optimizer`, `criterion` in basic pruner, replace `finetuner`, `evaluator` (just an evaluation function), `dummy_input` in iterative pruner.
3. preparation for next new future, add pruning target concepts.
4. add tests for most pruner, both old & new API.
5. bugfix: 
a. dependency aware mode sometimes invalid in iterative pruner.
b. add `best_result_mode` API for `TaskGenerator`, support the situation that some pruning results are not sorted according to evaluation metric

#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py [hunk 1]
```diff
@@ -425,11 +466,22 @@ class TaskGenerator:
         The log directory use to saving the task generator log.
     keep_intermediate_result
         If keeping the intermediate result, including intermediate model and masks during each iteration.
+    best_result_mode
+        The way to decide which one is the best result. Three modes are supported.
+        If the task results don't contain scores (task_result.score is None), it will fall back to ``latest``.
+
+        1. latest: The newest received result is the best result.
+        2. maximize_score: The one with largest task result score is the best result.
```
Comment (ultmaster) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:474]: Why `score`? Seems `maximize` and `minimize` are clear enough and aligned with HPO.

File: nni/algorithms/compression/v2/pytorch/utils/docstring.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+
+_EVALUATOR_DOCSTRING = r"""NNI will use the evaluator to intervene in the model training process, so as to perform training-aware model compression.
+        All training-aware model compression will use the evaluator as the entry for intervention training in the future.
+        Usually you just need to wrap some classes with ``nni.trace`` or package the training process as a function to initialize the evaluator.
+        Please refer ... for a full tutorial on how to initialize a ``evaluator``.
```
Comment (ultmaster) [inline nni/algorithms/compression/v2/pytorch/utils/docstring.py:8]: What is ...

File: nni/algorithms/compression/v2/pytorch/utils/docstring.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+        The following are two simple examples, if you use pytorch_lightning, please refer to LightningEvaluator,
```
Comment (ultmaster) [inline nni/algorithms/compression/v2/pytorch/utils/docstring.py:10]: Use ``:class:`~xxx.xxx.LightningEvaluator` ``.

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py [hunk 4]
```diff
@@ -126,24 +144,79 @@ def compress(self) -> Tuple[Module, Dict]:
...
+        # for fake __init__ overload, parsing args and kwargs, initializing evaluator or [trainer, traced_optimizer, criterion],
+        # return the remaining arguments.
+        if (len(args) > 0 and isinstance(args[0], Evaluator)) or (len(args) == 0 and isinstance(kwargs.get('evaluator', None), Evaluator)):
+            init_kwargs = self._parse_args(new_api, args, kwargs, init_kwargs)
+            self.evaluator: LightningEvaluator | TorchEvaluator = init_kwargs.pop('evaluator')
+            if not self.evaluator._initialization_complete:
+                self.evaluator._init_optimizer_helpers(model)  # type: ignore
+            self.using_evaluator = True
+        else:
+            init_kwargs = self._parse_args(old_api, args, kwargs, init_kwargs)
```
Comment (ultmaster) [inline nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:170]: Throw a deprecation warning here.

File: test/algo/compression/assets/simple_mnist/simple_torch_model.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.9, weight_decay=5e-4)
+    training_model(model, optimizer, F.nll_loss)
+
+
+def evaluating_model(model: Module, device: torch.device = device):
+    model.eval()
+
+    # prepare data
+    data_dir = Path(__file__).parent / 'data'
+    MNIST(data_dir, train=False, download=True)
```
Comment (ultmaster) [inline test/algo/compression/assets/simple_mnist/simple_torch_model.py:81]: We have a MNIST dataset in test/data. Check `pipelines/templates/download-test-data.yml`.

File: nni/algorithms/compression/v2/pytorch/pruning/tools/data_collector.py [hunk 6]
```diff
@@ -6,13 +6,16 @@
...
+from .base import TrainerBasedDataCollector
 
 _logger = logging.getLogger(__name__)
 
-__all__ = ['WeightDataCollector', 'WeightTrainerBasedDataCollector', 'SingleHookTrainerBasedDataCollector']
+__all__ = ['TargetDataCollector', 'EvaluatorBasedTargetDataCollector', 'EvaluatorBasedHookDataCollector',
+           'WeightDataCollector', 'WeightTrainerBasedDataCollector', 'SingleHookTrainerBasedDataCollector']  # TODO: remove in nni v3.0.
 
 
+# TODO: remove in nni v3.0.
```
Comment (ultmaster) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/data_collector.py:18]: Please throw a deprecation warning if users use this.

Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/data_collector.py:18]: most user will not use these classes, they are pruner utils, I think pruner level deprecation warning is enough, or we will have too many warning if user use the old Pruner API.

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py (hunk 4 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:170]: fixed

File: nni/algorithms/compression/v2/pytorch/utils/docstring.py (hunk 3 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/docstring.py]: fixed

File: nni/algorithms/compression/v2/pytorch/utils/docstring.py (hunk 2 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/utils/docstring.py:8]: a placeholder waiting for tutorial link...

**Review by ultmaster — APPROVED** [review]

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py [hunk 7]
```diff
@@ -104,54 +101,50 @@ class AutoCompressPruner(IterativePruner):
...
+    Notes
+    -----
     The full script can be found :githublink:`here <examples/model_compress/pruning/auto_compress_pruner.py>`.
-    """
+    """.format(evaluator_docstring=_EVALUATOR_DOCSTRING)
+
+    @overload
+    def __init__(self, model: Module, config_list: List[Dict], total_iteration: int, admm_params: Dict,
+                 sa_params: Dict, log_dir: str = '.', keep_intermediate_result: bool = False,
+                 evaluator: LightningEvaluator | TorchEvaluator | None = None, speedup: bool = False):
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:127]: it is a little strange that sometimes you put `evaluator` right after `config_list`, sometimes you put it at the end of the arg list.

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py [hunk 8]
```diff
@@ -252,6 +254,32 @@ def _remove_all_hook(self):
                 self._remove_hook(hook_id)
 
 
+class EvaluatorBasedDataCollector(DataCollector):
+    # TODO: add docstring
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:258]: why not add docstring in this pr?

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py (hunk 7 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:127]: yes. I prefer `model, config_list, evaluator`, the reason why it place everywhere is for aligning with old API, after deprecating the old API, I will change them all.

**Review by QuanluZhang — APPROVED** [review]
